File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    answer = number[0]
    cnt = 0
    for i, n in enumerate(number[1:], 1):
        if n <= answer[-1] or cnt >= k:
            answer += n
        else:
            j = len(answer) -1
            while  max(0, len(answer)-k+cnt) <= j -1 and answer[j-1] < n:
                j -= 1
            cnt += len(answer[max(0, j):])
            answer = answer[:max(0, j)] + n
    return answer[:len(number)-k]

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.Arrays;
import java.util.ArrayList;

class Solution {
	    public static String solution(String number, int k) {
	        ArrayList<Character> al = new ArrayList(number.length() - k);
	        int index = 0;
	        int cnt = 0;
	        al.add(number.charAt(0));
	        for (int i = 1; i < number.length(); i++) {
	        	char c = number.charAt(i);
	            if (al.get(index) >= c || cnt >= k) {
	                al.add(c);
	                index++;
	            } else {
	                while (al.size() > 0 && cnt < k && al.get(index) < c) {
	                    cnt++;
	                    al.remove(index--);
	                }
	                al.add(c);
	                index++;
	            }
	        }
            
            String answer = "";
            for (int i = 0; i < number.length() - k; i++) answer += al.get(i);
            return answer;
	    }
	}